### PR TITLE
SQL: non ISO 8601 versions of DAY_OF_WEEK and WEEK_OF_YEAR functions

### PR DESCRIPTION
**Pull request by astefan**
Renamed DAY_OF_WEEK and WEEK_OF_YEAR functions to their ISO version and added the same functions with different functionality. Rewritten the datetime functions documentation to follow the format of the other functions documentation pages.

Fixes https://github.com/elastic/elasticsearch/issues/36263

### CODE REVIEW

**Review by costin — APPROVED** [review]
LGTM but I would wait for @matriv's PR that fixes the naming convention inside `FunctionRegistry`.

**Comment by astefan** [comment]
run the gradle build tests 1

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/NonISODateTimeProcessor.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+public class NonISODateTimeProcessor extends BaseDateTimeProcessor {
+    
+    public enum NonISODateTimeExtractor {
+        DAY_OF_WEEK(zdt -> {
+            // by ISO 8601 standard, Monday is the first day of the week and has the value 1
+            // non-ISO 8601 standard considers Sunday as the first day of the week and value 1
+            int dayOfWeek = zdt.get(ChronoField.DAY_OF_WEEK) + 1;
+            return dayOfWeek == 8 ? 1 : dayOfWeek;
+        }),
+        WEEK_OF_YEAR(zdt -> {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/NonISODateTimeProcessor.java]: Could you add here also the difference between iso and non-iso?

File: x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt [hunk 2]
```diff
@@ -98,8 +98,10 @@ class org.elasticsearch.xpack.sql.expression.function.scalar.whitelist.InternalS
 #
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt]: Seems we are missing the existing non-iso methods here, and relevant tests that would catch those cases.

File: docs/reference/sql/functions/date-time.asciidoc [hunk 3]
```diff
@@ -3,83 +3,388 @@
...
+*Input*:
+
+<1> datetime function name
+<2> date expression
+
+*Output*: integer
+
+.Description:
+
+Extract fields from datetimes by specifying the name of a <<sql-functions-datetime,datetime function>>.
```
Comment (matriv) [inline docs/reference/sql/functions/date-time.asciidoc]: I would rephrase: ... `from a datetime field` or so, I don't like the datetime`s` plural so much :-)

**Review by matriv — CHANGES_REQUESTED** [review]
Looks good! Left some comments, and will wait for the missing methods in painless.

File: x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt (hunk 2 above)
Comment (astefan) [reply x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt]: The iso methods are handled through the `dateTimeChrono` method which also handles the rest of the date time functions. I've renamed the `DateTimeExtractor` enum values for ISO day_of_week and week_of_year functions, hopefully making this less confusing.

**Review by matriv — APPROVED** [review]
LGTM! You need to adjust the function registration on top of the change with mandatory primary name.

**Comment by astefan** [comment]
@elasticmachine run default distro tests
